modules/account-lib/src/coin/sol/solInstructionFactory.ts

```diff
@@ -0,0 +1,108 @@
+import { InstructionBuilderTypes, MEMO_PROGRAM_PK } from './constants';
+import { InstructionParams, Memo, Nonce, Transfer, WalletInit } from './iface';
+
+/**
+ * Construct Solana instructions from instructions params
+ *
+ * @param {InstructionParams} instructionToBuild - the data containing the instruction params
+ * @returns {TransactionInstruction[]} An array containing supported Solana instructions
+ */
+export function solInstructionFactory(instructionToBuild: InstructionParams): TransactionInstruction[] {
+  switch (instructionToBuild.type) {
+    case InstructionBuilderTypes.NonceAdvance:
+      return advanceNonceInstruction(instructionToBuild);
+    case InstructionBuilderTypes.Memo:
+      return memoInstruction(instructionToBuild);
+    case InstructionBuilderTypes.Transfer:
+      return transferInstruction(instructionToBuild);
+    case InstructionBuilderTypes.CreateNonceAccount:
+      return createNonceAccountInstruction(instructionToBuild);
+    default:
+      throw new Error(`Invalid instruction type or not supported`);
+  }
+}
+
+/**
+ * Construct Advance Nonce Solana instructions
+ *
+ * @param {Nonce} data - the data to build the instruction
+ * @returns {TransactionInstruction[]} An array containing Advance Nonce Solana instruction
+ */
+function advanceNonceInstruction(data: Nonce): TransactionInstruction[] {
+  const {
+    params: { authWalletPubKey, walletNoncePubKey },
+  } = data;
+  assert(authWalletPubKey, 'Missing authWalletPubKey param');
+  assert(walletNoncePubKey, 'Missing authWalletPubKey param');
+  const nonceInstruction = SystemProgram.nonceAdvance({
+    noncePubkey: new PublicKey(walletNoncePubKey),
+    authorizedPubkey: new PublicKey(authWalletPubKey),
+  });
+  return [nonceInstruction];
+}
+
+/**
+ * Construct Memo Solana instructions
+ *
+ * @param {Memo} data - the data to build the instruction
+ * @returns {TransactionInstruction[]} An array containing Memo Solana instruction
+ */
+function memoInstruction(data: Memo): TransactionInstruction[] {
+  const {
+    params: { memo },
+  } = data;
+  assert(memo, 'Missing memo param');
+  const memoInstruction = new TransactionInstruction({
+    keys: [],
+    programId: new PublicKey(MEMO_PROGRAM_PK),
+    data: Buffer.from(memo),
+  });
+  return [memoInstruction];
+}
+
+/**
+ * Construct Transfer Solana instructions
+ *
+ * @param {Transfer} data - the data to build the instruction
+ * @returns {TransactionInstruction[]} An array containing Transfer Solana instruction
+ */
+function transferInstruction(data: Transfer): TransactionInstruction[] {
+  const {
+    params: { fromAddress, toAddress, amount },
+  } = data;
+  assert(fromAddress, 'Missing fromAddress param');
+  assert(toAddress, 'Missing toAddress param');
+  assert(amount, 'Missing toAddress param');
+  const transferInstruction = SystemProgram.transfer({
+    fromPubkey: new PublicKey(fromAddress),
+    toPubkey: new PublicKey(toAddress),
+    lamports: new BigNumber(amount).toNumber(),
```

Comment:
Seems to be ok:
https://github.com/solana-labs/solana-web3.js/blob/master/src/system-program.ts#L36